Update idle loop to reduce calls to suspend

c1728p9 opened the pull request:
Change tickless handling to use public RTX calls and to prevent unnecessary calls to suspend/resume by looping until the kernel needs to be resumed.

c1728p9 committed:
Update idle loop to reduce calls to suspend

Change tickless handling to use public RTX calls and to prevent
unnecessary calls to suspend/resume by looping until the kernel
needs to be resumed.

c1728p9 commented:
This change also allows #6536 to work correctly, as the low power timer no longer needs to written to on every wakeup.

c1728p9 commented:
CC @JonatanAntoni @bulislaw

kjbracey-arm commented on rtos/TARGET_CORTEX/mbed_rtx_idle.cpp:
```diff
@@ -95,21 +95,22 @@ uint32_t OS_Tick_GetInterval (void) {
+        if (!osRtxInfo.kernel.pendSV) { // check devices didn't run any interrupt handlers kicking the OS between suspend and critical section
+            sleep();                    // uses WFI - will wake as soon as any pending interrupts
+        }
+        core_util_critical_section_exit();
```
There might be call for an ISB here - you're wanting pending interrupts to trigger between this line and the next one, and an ISB should make sure that happens.

In theory I think you could redisable interrupts before any have a chance to trigger - formally the sequence "enable IRQs / disable IRQs" needs an ISB somewhere in the middle to guarantee they have time to trigger.

Relevant ARM ARM quote:

> If execution of a CPS instruction:
> * Increases the execution priority, the CPS execution serializes that change to the instruction stream.
> * Decreases the execution priority, the architecture guarantees only that the new priority is visible to instructions executed after either executing an ISB, or performing an exception entry or exception return.

(In practice they're likely to trigger before you hit the next disable, which means you'd see pendSV set by the next time you hit line 104.)

kjbracey-arm commented on rtos/TARGET_CORTEX/mbed_rtx_idle.cpp:
```diff
@@ -95,21 +95,22 @@ uint32_t OS_Tick_GetInterval (void) {
-        os_timer->cancel_tick();
-        // calculate how long we slept
-        elapsed_ticks = os_timer->update_tick();
+        if (osRtxInfo.kernel.pendSV) {
```
This could just be part of the while condition? Could be true on first entry to the while

JonatanAntoni commented on rtos/TARGET_CORTEX/mbed_rtx_idle.cpp:
```diff
@@ -95,21 +95,24 @@ uint32_t OS_Tick_GetInterval (void) {
+    while (!os_timer->suspend_time_passed() && !event_pending) {
+
+        core_util_critical_section_enter();
+        if (osRtxInfo.kernel.pendSV) {
```
osRtxInfo.kernel.pendSV is not public API. Perhaps @RobertRostohar can justify if this solution is solid and a small API change would be valuable.

0xc0170 replied:
Might help with the review, details are here https://github.com/ARMmbed/mbed-os/pull/6273#issuecomment-375278530

kjbracey-arm replied:
Agree that there should be better way - pendSV peeking was the best I could think of with current APIs.

The "simple" API change I would suggest is to permit osKernelSuspend/Resume to be callable from critical section (which is what happens via direct svcKernelSuspend calls at the moment, and this replaces), or to have them or equivalents actually return from the suspend call such that the thread is now in a critical section.

I guess in general one also needs to think about priviliged/unprivileged operation, in case a system is using User mode - system suspension, critical sections and the WFI instruction are privileged operations anyway, so this whole "normal thread code does suspend" only works if the idle thread is privileged (or the suspend call makes you privileged?), or you have a WFE-only version (which I believe is less flexible and can reduce power saving opportunities).

c1728p9 replied:
@JonatanAntoni when I was initially working on tickless I proposed on extension to the RTOS to allow this, which you might be able to use:
https://github.com/ARMmbed/mbed-os/pull/5031/files#diff-415fd494349b5414064e0cc0d9f0d596R189

As for this PR, how do you want to proceed? Can we move forward using this internal API until RTX has a public API which supports this?

JonatanAntoni replied:
Sure, for the time being you can access this internal bit. But as it is not public API it might change in the future without being warned. So we should align the low power operation, soon. We need to address this privileged/unprivileged issue nonetheless talking about MPU protection.

c1728p9 commented:
/morph build

0xc0170 commented:
`mbedmicro-rtos-mbed` needs an update?

cmonr commented:
@0xc0170 Would appear to be the case.

c1728p9 committed:
Update the SysTimer test

Update the SysTimer test to match the updated API. Changes are:
- increment_tick() renamed to _increment_tick() and explicitly
    synchronized
-update_tick() replaced with resume() and a call to suspend() was added
    before this

c1728p9 commented:
Even with the test update, this PR intermittently [fails on test_update_tick](https://github.com/ARMmbed/mbed-os/pull/6534/files#diff-ed008b41e4b77db2ce117cdd27e5745cR106) on the nrf51. From profiling this test, I found that the call to suspend is taking at times +500us. I suspect that a tick is passing and that is why 1 is returned from suspend rather than 0 - `FAIL: Expected 0 Was 1`.

If I pull in https://github.com/ARMmbed/mbed-os/pull/5629 suspend takes around 133us to complete, which is still a lot, but the test consistently passes. @fkjagodzinski you were the author of this test right? Do you have any recommendations on how I can get this test to pass?

c1728p9 commented:
/morph build

fkjagodzinski commented:
@c1728p9 `test_update_tick()` is supposed to check just the `update_tick()` method with, and without a delay in between calls. Since `update_tick()` functionality was moved to `resume()` it's enough to update method name in the test. No calls to `suspend()` needed.

With this patch 5bfaf528d571de6ae67fe50c21ff25af9bb6fecf all `tests-mbedmicro-rtos-mbed-systimer` consistently pass on NRF51_DK.

0xc0170 commented:
> With this patch 5bfaf52 all tests-mbedmicro-rtos-mbed-systimer consistently pass on NRF51_DK.

Shall this be added here?

cmonr commented:
@fkjagodzinski Interesting. Did that commit end up going into a feature branch? (Not the simplest to tell with just the sha)

I would suggest opening a new PR to bring this into master.

fkjagodzinski commented:
> (Not the simplest to tell with just the sha)

Sorry for that; this commit is from my fork of mbed-os https://github.com/fkjagodzinski/mbed-os/commit/5bfaf528d571de6ae67fe50c21ff25af9bb6fecf ; it is not present in any PR. I tested it locally with NRF51_DK.

c1728p9 commented:
@fkjagodzinski I'm not sure if it makes sense to call `resume` without first calling `suspend`. It works now, but this might not be the case if the code is refactored in the future.

Could the test be updated to use a mock timer rather than the low power timer? This would allow more targeted and deterministic testing. It also would not be intrusive to the running system. Also, the latency being implicitly tested in this test is being explicitly tested for in the branch feature-hal-spec-ticker.

fkjagodzinski committed:
RTOS: SysTimer: Update tests

Use a mock ticker object instead of the lp_ticker for update_tick() and get_time() tests.

fkjagodzinski commented:
In https://github.com/fkjagodzinski/mbed-os/commit/20013d75d6b09fc2fb64b1d11dc1ec7bf3f4ed53 I've updated two test cases to use a mock ticker object instead of the lp_ticker. The changes are prepared on top of this PR, so could be cherry-picked here.

c1728p9 commented:
I pulled in @fkjagodzinski's change for testing.

cmonr commented:
/morph build

fkjagodzinski committed:
RTOS: SysTimer: Fix test compilation error on ARM & IAR

fkjagodzinski commented:
My fault -- I tested that patch with GCC toolchain only. This commit https://github.com/fkjagodzinski/mbed-os/commit/b5dc7f437beaafe973e85c34c877e46165e69bc6 fixes ARM and IAR compilation errors.

c1728p9 commented:
Thanks for the fix @fkjagodzinski. I pulled it into this PR.

cmonr commented:
/morph build

0xc0170 commented:
/morph build

0xc0170 commented:
Ah too slow to remove the comment to trigger another build :-) I invoked build instead of test to retest 😖

fkjagodzinski commented:
`tests-mbed_drivers-timeout` / `'Multiple timeouts running in parallel (attach)'`test case compiled with ARM toolchain failed again on NRF52_DK board.

I'm investigating. So far, I can tell that the issue is also present on current master branch. #6515 might help here -- with this PR merged I couldn't reproduce the issue locally any more.

0xc0170 commented:
> I'm investigating. So far, I can tell that the issue is also present on current master branch. #6515 might help here -- with this PR merged I couldn't reproduce the issue locally any more.

Interesting, and that fix is blocked by another failure 😕

fkjagodzinski commented:
@0xc0170 No, not until I saw morph failures for this PR. It didn't look related to this PR, that's why I checked master. The issue does not show up very often, but can be reproduced.

fkjagodzinski commented:
I've created an issue #6732 based on the failing test case `tests-mbed_drivers-timeout` / `'Multiple timeouts running in parallel (attach)'`.

stevew817 commented on rtos/TARGET_CORTEX/SysTimer.cpp:
```diff
@@ -61,23 +70,30 @@ void SysTimer::setup_irq()
-void SysTimer::cancel_tick()
-{
-    remove();
+    schedule_tick(ticks);
```
@c1728p9 Why is there no remove() call here? I'm seeing issues on my system because of this.

The SysTimer is based on a TimerEvent class, which has a static member for the timer's event data struct. Additionally, the call to suspend may happen before the previously set timer expires (for example when all threads go idle and the idle handler gets activated). At that point, the timer event data is still present in the timer queue. Calling schedule_tick here will insert *the same* event structure again in the queue. And since the queue is by reference, that'll lead to a recursive timer that breaks the system.

c1728p9 replied:
Hi @stevew817 I gave this a try but was unable to reproduce the failure. Do you have example code which reproduces the issue?

From both code analysis and running the code can I cant fix a sequence which causes this. The basic flow of the idle loop should unconditionally call remove before each insert as indicated below:
```
while (true) {
    osKernelSuspend();      // removes event
    os_timer->suspend();    // inserts event
    sleep();
    os_timer->resume();     // removes event
    osKernelResume();       // inserts event
}
```

With a debugger I can confirm the following sequence is being met by setting a breakpoint in TimerEvent::remove, TimerEvent::insert and TimerEvent::insert_absolute. Below is the callstack from each:
```
osKernelSuspend()
    svcRtxKernelSuspend()
        KernelBlock()
            OS_Tick_Disable()
                os_timer->cancel_tick()
                    remove()

os_timer->suspend()
    schedule_tick()
        insert_absolute()

os_timer->resume()
    remove()

osKernelResume()
    svcRtxKernelResume()
        KernelUnblock()
            OS_Tick_Enable()
                schedule_tick()
                    insert_absolute();
```

stevew817 replied:
I ran across this when running cloud-client-example using Thread as a sleepy node in tickless mode. The code would hang, and attaching a debugger confirmed it was stuck in the tick handler IRQ because somehow the tick event had managed to add itself recursively (the next pointer of the tick event data was pointing to the same event data, causing the remove call to not do anything since the head just always ands up being the tick event data).

I fixed the issue by adding a cancel_tick() call before the schedule_ticks call in SysTimer::suspend. I didn't investigate further after solving my issue, so I'm afraid I don't have a reproducer ready.

cmonr commented:
@c1728p9 Please take a look at this and if this needs a fix and/or update, we can take a fix in for RC2.

CC: @ARMmbed/mbed-os-maintainers
